### DISCUSSION TITLE
How to deal with serializing jobs that are dependent on unserializable objects like sockets/network connections?

### DISCUSSION THREAD

**Discussion by DonDebonair** [agronholm/apscheduler#656] [Q&A]
Hey folks! I posted this question on Gitter, but @agronholm suggested I posted it on Github Discussions because my question fits the Q/A format better. I think he's right.

I have a question on best practices for persisting jobs that require external network/socket connections. Imagine the following: 

- You have bunch of jobs you want to schedule in APScheduler and these jobs take the form of methods on a class instance
- The class in question has a field that holds a socket/network connection of some kind. For example: a websocket connection to Slack

You obviously cannot put these jobs in a persistent job store, because sockets are not serializable. How would you deal with this? I can think of 2 solutions, both of which are suboptimal:

- You define `__getstate__`for serialization on the class that holds the job function and leave out the field with the socket. Then you also define `__setstate__` for deserialization and make sure to instantiate a new socket. The big drawbacks seems to me 2-fold:
    - Your connection/socket could hold some kind of state accrued over the lifetime of your application, and you cannot easily recover this state
    - You create a sort of memory-leak, because new sockets are created anytime a job is deserialized and run
- You make your socket/connection into a Singleton of which there will never exist more than one and the class with the job functions doesn’t hold a field with the socket/connection anymore, but instead always calls `ConnectionCls.get_instance()` to get the connection for whatever it wants to do. This way, only one instance of the connection will ever exist. Huge drawback (to me anyways): Singletons feel like a huge code smell, because you cannot easily pass connection instances around anymore and it makes code hard to test

Am I missing a third alternative? Does anyone have experience with this kind of situation?

**Comment by agronholm** [comment]
So let's say you have an instance of that class, and you want to schedule it to do something with sockets.

One alternative I can think of right off the bat is to use the scheduled job to look up that object, possibly importing it from another module, and then call the appropriate method on that object. If you need this to work dynamically, then you can just give the attribute name of the object in the module where it is to be found, as an argument to the scheduled job. Alternatively you can store all named instances of the class in a dictionary, available on the module level, and just look up the appropriate object using the argument passed to the job.

**Reply by DonDebonair** [reply]
I don't know if your suggestion would work exactly. To clarify my problem, I'll elaborate a bit on what I'm building. I'm building a plugin-framework that can be used to build Slack bots. How it works, is as follows:

1. Developers write plugins by subclassing a class that I provide. Let's call it `SlackBasePlugin`. This base class contains all kinds of convenience methods to post messages to Slack, add reactions, upload files etc.
2. When implementing a plugin class, developers use decorators that my library provides to make methods in their class run when certain things happen. Examples:
	- `@listen_to(r"a message")` will trigger the decorated method when `a message` is posted on Slack
	- `@process("reaction_added")` will trigger the decorated method when the `reaction_added` event happens in Slack
	- `@schedule("* * */5")` will trigger the decorated method on the specified schedule -> this decorator is powered by APScheduler
    Why are developers forced to subclass and write methods and not regular functions? That is so that developers can use all the 	 
    conviences that the base class provides
3. Users/developers provide the import paths to plugins in a configuration file for the bot they want to run. These plugins are then dynamically loaded and instantiated on startup.
4. In order for the plugin classes to have access to all the Slack functionality that my plugin framework provides (through the base class), a _slack client_ instance is passed into the constructor when instantiating each plugin. So basically:

```python
plugin = imported_cls(slack_client, settings)
plugin_registry.register(plugin)
```

Mind you, this slack client is quite low-level and not supposed to be used directly by plugin developers. Instead, convenience methods on the base class wrap low-level calls to the `slack_client`. It is this `slack_client` object that makes it hard to pickle the plugin class

Any method decorated with `@schedule` will be scheduled in APScheduler and the class instance of which the method is a part, is passed as argument (`self`) to the job.

I'm not sure how your solution would work in this scenario. Maybe you haven an idea?

Thanks for the help! And if you feel this takes too much of your time, I also understand. So feel free to politely blow me off :)

btw, [this](https://github.com/DonDebonair/slack-machine) is the project I'm working on. Currently, it uses the Singleton solution I mentioned before. But I'm working on a rewrite to move to asyncio, and I want to take that opportunity to remove as many warts from the code as possible

And [here](https://github.com/DonDebonair/sm-example-plugin/blob/main/example_plugin/my_plugin_async.py) you can find an example of a plugin. It doesn't have any `@schedule` methods yet

**Reply by agronholm** [reply]
This wouldn't be a problem with the memory store, but since persistent job stores are involved, this gets a tad more complicated since the arguments to the job (`self` most importantly) cannot be serialized. The only way then is to look up the necessary objects using module level variables or functions in the scheduled function (a wrapper perhaps?). Your project looked pretty complex so I'm not going to dive very deep into it but if you can simplify the use case enough for me, and show a way to gain access to the appropriate method instances, then I can try to give you an example.

**Reply by DonDebonair** [reply]
Hey @agronholm! Sorry for not responding sooner. I've been thinking it over a bit, and I find it hard to give you a simple example that you can actually run.

I think you've set me on the right track with the module variables. Modules are initialized once in Python as far as I understand, so if I have a Python module that creates a SlackClient instance in a global variable, then I could import that whever. It feels nearly as ugly as the Singleton pattern, but I don't see another way 😅
